PerspectiveRailroad/PerspectiveRailroad.js

```diff
@@ -0,0 +1,302 @@
+  const left_endpoints = [point_a, second_line.meet(guide_left)];
+  const right_endpoints = [point_b, second_line.meet(guide_right)];
+
+  // Find the center of the rectangle, then draw a line through it and the
+  // vanishing point.
+  const diag1 = left_endpoints[1].join(point_b);
+  const diag2 = right_endpoints[1].join(point_a);
+  const center = diag1.meet(diag2);
+  const center_line = center.join(vp);
+
+  for (let i = 0; i < MAX_ITERATIONS; i++) {
+    const prev_left = left_endpoints[left_endpoints.length - 2];
+    const current_line = horizontal_lines[horizontal_lines.length - 1];
+    const center = current_line.meet(center_line);
+    const line_to_next = prev_left.join(center);
+    const next_right = line_to_next.meet(guide_right);
+    const next_line = next_right.join(Point.DIR_X);
+    const next_left = next_line.meet(guide_left);
+
+    horizontal_lines.push(next_line);
+    left_endpoints.push(next_left);
+    right_endpoints.push(next_right);
+  }
+
+  return {
+    guide_left,
+    guide_right,
+    horizontal_lines,
+    left_endpoints,
+    right_endpoints,
+    vanishing_point: vp,
+  };
+}
+
+function railroad_ties(tie_bottoms, tie_thickness) {
+  const ties = [];
+
+  const point_a = tie_bottoms.left_endpoints[0];
+  const point_b = tie_bottoms.right_endpoints[1];
+  const ref_point = Point.lerp(point_a, point_b, tie_thickness);
+  const ref_line = ref_point.join(tie_bottoms.vanishing_point);
+
+  const { guide_left, guide_right } = tie_bottoms;
+  const quad_count = tie_bottoms.horizontal_lines.length - 1;
+  for (let i = 0; i < quad_count; i++) {
+    const quad_a = tie_bottoms.left_endpoints[i];
+    const quad_b = tie_bottoms.right_endpoints[i];
+    const quad_c = tie_bottoms.right_endpoints[i + 1];
+
+    const diag = quad_a.join(quad_c);
+    const isx = diag.meet(ref_line);
+    const tie_top = isx.join(Point.DIR_X);
+    const top_left = tie_top.meet(guide_left);
+    const top_right = tie_top.meet(guide_right);
+
+    ties.push(new PolygonPrimitive([quad_a, quad_b, top_right, top_left]));
+  }
+
+  return ties;
+}
+
+function make_background() {
+  const sky_prim = new RectPrimitive(
+    Point.point(0, 0),
+    Point.direction(WIDTH, VP_RAILS.y)
+  );
+  const sky = new GroupPrimitive([sky_prim], SKY_STYLE);
+
+  const ground_prim = new RectPrimitive(
+    Point.point(0, VP_RAILS.y),
+    Point.direction(WIDTH, HEIGHT - VP_RAILS.y)
+  );
+  const ground = new GroupPrimitive([ground_prim], GROUND_STYLE);
+
+  return new GroupPrimitive([sky, ground]);
+}
+
+class AnimatedTie {
+  constructor(quad, start_frame, duration_frames) {
+    if (quad.vertices.length != 4) {
+      throw new Error("quad must have exactly 4 vertices");
+    }
+
+    this.quad = quad;
+    this.start_frame = start_frame;
+    this.duration_frames = duration_frames;
+  }
+
+  compute_polygon(frame) {
+    // We haven't started animating, so don't render anything
+    if (frame < this.start_frame) {
```

Comment:
In time I might want to make a tweening object for this. Let's see how much I animate like this.